File: core/r/lib/io.R
# Reading tidy data and checksumming inputs.

# Read a tidy CSV or Excel file into a data frame.
read_tidy <- function(path, sheet = NULL) {
  if (!file.exists(path)) stop(sprintf("input file not found: %s", path))
  ext <- tolower(tools::file_ext(path))
  if (ext %in% c("xlsx", "xls")) {
    as.data.frame(readxl::read_excel(path, sheet = sheet %||% 1))
  } else {
    as.data.frame(readr::read_csv(path, show_col_types = FALSE, progress = FALSE))
  }
}

# md5 and sha256 of the exact bytes of a file.
file_checksums <- function(path) {
  list(
    md5    = unname(tools::md5sum(path)),
    sha256 = digest::digest(file = path, algo = "sha256")
  )
}

File: core/python/pubplot/io.py
"""Reading tidy data and checksumming inputs (mirrors the R io helpers)."""
import hashlib
import os

import pandas as pd


def read_tidy(path, sheet=None):
    if not os.path.exists(path):
        raise FileNotFoundError(f"input file not found: {path}")
    ext = os.path.splitext(path)[1].lower()
    if ext in (".xlsx", ".xls"):
        return pd.read_excel(path, sheet_name=0 if sheet is None else sheet)
    return pd.read_csv(path)


def file_checksums(path):
    md5, sha = hashlib.md5(), hashlib.sha256()
    with open(path, "rb") as f:
        for chunk in iter(lambda: f.read(65536), b""):
            md5.update(chunk)
            sha.update(chunk)
    return {"md5": md5.hexdigest(), "sha256": sha.hexdigest()}
